cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -878,67 +847,86 @@ func (p *BicepProvider) modulePath() string {
 func (p *BicepProvider) ensureParameters(
 	ctx context.Context,
 	asyncContext *async.InteractiveTaskContextWithProgress[*DeploymentPlan, *DeploymentPlanningProgress],
-	deployment *Deployment,
-) error {
-	if len(deployment.Parameters) == 0 {
-		return nil
+	template azure.ArmTemplate,
+	parameters azure.ArmParameters,
+) (azure.ArmParameters, error) {
+	if len(template.Parameters) == 0 {
+		return azure.ArmParameters{}, nil
 	}
 
-	for key, param := range deployment.Parameters {
-		// If this parameter has a default, then there is no need for us to configure it
-		if param.HasDefaultValue() {
+	configuredParameters := make(azure.ArmParameters, len(template.Parameters))
+
+	sortedKeys := maps.Keys(template.Parameters)
+	slices.Sort(sortedKeys)
+
+	configModified := false
+
+	for _, key := range sortedKeys {
+		param := template.Parameters[key]
+
+		// If a value is explicitly configured via a parameters file, use it.
+		if v, has := parameters[key]; has {
+			configuredParameters[key] = v
 			continue
 		}
-		if !param.HasValue() {
-			configKey := fmt.Sprintf("infra.parameters.%s", key)
 
-			if v, has := p.env.Config.Get(configKey); has {
-				param.Value = v
-				deployment.Parameters[key] = param
-				continue
+		// If this parameter has a default, then there is no need for us to configure it.
+		if param.DefaultValue != nil {
+			continue
+		}
+
+		// This required parameter was not in parameters file - see if we stored a value in config from an earlier
+		// prompt and if so use it.
+		configKey := fmt.Sprintf("infra.parameters.%s", key)
+
+		if v, has := p.env.Config.Get(configKey); has {
+			configuredParameters[key] = azure.ArmParameterValue{
+				Value: v,
```

Comment:
We could validate for type-mismatch here. Reading through the code, `armParameters` eventually is marshaled into a `json` string by `deploymentClient` that is then sent as part of the deployment request. I expect that this currently just fails the server deployment validation request late in the cycle.

Reply:
I'll add a type test here and if the types don't match up discard the saved value (if there is one) and then just do the prompting.